internal/controller/watcher_controller.go

```diff
@@ -144,7 +144,14 @@ func (r *WatcherReconciler) handleDeletingState(ctx context.Context, watcherCR *
 func (r *WatcherReconciler) handleProcessingState(ctx context.Context,
 	watcherCR *v1beta2.Watcher,
 ) (ctrl.Result, error) {
-	virtualSvc, err := r.IstioClient.NewVirtualService(ctx, watcherCR, r.WatcherVSNamespace)
+	gateways, err := r.IstioClient.ListGatewaysByLabelSelector(ctx, &watcherCR.Spec.Gateway.LabelSelector)
+	if err != nil || len(gateways.Items) == 0 {
+		r.EventRecorder.Event(watcherCR, "Warning", "WatcherGatewayNotFound",
+			"Watcher: Gateway for the VirtualService not found")
+		return r.updateWatcherState(ctx, watcherCR, shared.StateError, err)
+	}
```

Comment:
I have moved the length check and event recording outside of the client. If there is no error in the client, it shouldn't need to worry about if there are objects or not. The caller needs to determine if an empty list is an error case or not. In this case yes, since this is the behavior we had before.